### PR TITLE
feat(graph): Transaction Cache Engine for XRP (in-memory only)

### PR DESCRIPTION
**Pull request by Nodonisko**
In-memory caching for transactions so we can do some computations that are needed for graph. In future once I introduce permanent storage we can replace some usages of TrezorConnect.getAccountInfo with this lib.

TODO:
- [x] XRP balance changes are much higher than actual balance - investigate

## Related Issue

Resolve

### CODE REVIEW

File: suite-native/accounts/src/TransactionCacheProvider.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { useEffect } from 'react';
+import { useSelector } from 'react-redux';
+
+import { TransactionCacheEngine } from '@suite-common/transaction-cache-engine';
+import { selectAccounts } from '@suite-common/wallet-core';
+
+export const TransactionCacheProvider: React.FC = () => {
```
Comment (juriczech) [inline suite-native/accounts/src/TransactionCacheProvider.ts]: nit: Seems like this could just be a hook?

File: suite-common/transaction-cache-engine/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
+/* eslint-disable require-await */
```
Comment (juriczech) [inline suite-common/transaction-cache-engine/src/index.ts]: Please remove unused code

File: suite-native/module-accounts-import/src/screens/XpubScanScreen.tsx [hunk 3]
```diff
@@ -76,29 +76,29 @@ export const XpubScanScreen = ({
...
-                    secondaryButtonTitle: 'Where to find it?',
-                    onPressSecondaryButton: () => {
-                        hideAlert();
-                        setIsHintSheetVisible(true);
-                    },
-                });
-            }, 1000);
-            return;
-        }
+        // if (
```
Comment (juriczech) [inline suite-native/module-accounts-import/src/screens/XpubScanScreen.tsx]: Please revert when finished

File: suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+            ) {
+                break;
+            }
+
+            if (networkType === 'ripple' && !result.payload.marker) {
+                break;
+            }
+        }
+
+        if (!(await this.accountExists({ coin, descriptor }))) {
```
Comment (juriczech) [inline suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts:165]: nit: I think it would be more readable if you stored the awaited value before, but might just be me

File: suite-common/transaction-cache-engine/src/tests/TransactionCacheEngine.test.ts [hunk 5]
```diff
@@ -0,0 +1,137 @@
+import TrezorConnect from '@trezor/connect';
+
+import { TransactionCacheEngine } from '../TransactionCacheEngine';
+import { MemoryStorage } from '../MemoryStorage';
+import { accountBalanceHistoryResult as btcAccountBalanceHistoryResult } from './__fixtures__/btc';
+import { AccountBalanceHistory, AccountUniqueParams } from '../types';
+import { accountBalanceHistoryResult as rippleAccountBalanceHistoryResult } from './__fixtures__/xrp';
+
+jest.mock('@trezor/connect', () => {
```
Comment (szymonlesisz) [inline suite-common/transaction-cache-engine/src/tests/TransactionCacheEngine.test.ts:9]: pls do not use custom connect mocks, use shared mock from test-utils,
changes example: https://github.com/trezor/trezor-suite/pull/10100

File: suite-common/wallet-config/src/networksConfig.ts [hunk 6]
```diff
@@ -1,6 +1,9 @@
-import { DeviceModelInternal } from '@trezor/connect';
 import type { ExtendedMessageDescriptor } from '@suite-common/intl-types';
 import type { Keys, Without } from '@trezor/type-utils';
+import { DeviceModelInternal } from '@trezor/connect';
+
+// DeviceModelInternal must be imported from @trezor/protobuf directly, otherwise it crashes tests in TransactionCacheEngine
```
Comment (szymonlesisz) [inline suite-common/wallet-config/src/networksConfig.ts]: pls rm leftovers

File: suite-common/graph/src/graphDataFetching.ts [hunk 7]
```diff
@@ -75,30 +112,32 @@ export const getAccountBalanceHistory = async ({
...
 
     const accountMovementHistoryWithBalance = addBalanceForAccountMovementHistory(
-        accountMovementHistory.payload,
+        accountMovementHistory,
         coin,
     );
 
     // Last point must be balance from getAccountInfo because blockchainGetAccountBalanceHistory it's not always reliable for coins like ETH.
-    // TODO: We can get value from redux store instead of fetching it again?
+    // TODO: We can get value from Transaction Cache engine instead of fetching it again?
```
Comment (juriczech) [inline suite-common/graph/src/graphDataFetching.ts:138]: You're already getting the value from `TransactionCacheEngine`, so I guess this comment is redundant?

File: suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+    getAccountUniqueKey,
+} from './utils/asyncUtils';
+import { AccountUniqueKey, AccountUniqueParams, TransactionCacheEngineStorage } from './types';
+import { getAccountBalanceHistory } from './utils/balanceHistory';
+
+const DEBUG_ENABLED = false;
+const debugLog = (...args: any[]) => {
+    if (DEBUG_ENABLED) {
+        // eslint-disable-next-line no-console
+        console.log(...args);
```
Comment (juriczech) [inline suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts:21]: Right now, this code is unreachable since DEBUG_ENABLED is always going to be false which makes all of the `debugLog` calls redundant?

File: suite-native/accounts/src/transactionCacheMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+            accountsActions.updateAccount.match(action) ||
+            accountsActions.createAccount.match(action)
+        ) {
+            if (action.payload.networkType === 'ripple') {
+                TransactionCacheEngine.addAccount({
+                    coin: action.payload.symbol,
+                    descriptor: action.payload.descriptor,
+                });
+            }
+        }
```
Comment (juriczech) [inline suite-native/accounts/src/transactionCacheMiddleware.ts]: Isn't this duplication of `suite-native/accounts/src/useTransactionCache.ts`. Maybe I'm missing something 🤔

If new account gets created, it will trigger useEffect in the hook and call the same method, no?

File: suite-native/module-accounts-import/src/screens/XpubScanScreen.tsx [hunk 10]
```diff
@@ -96,7 +97,7 @@ export const XpubScanScreen = ({
                         setIsHintSheetVisible(true);
                     },
                 });
-            }, 1000);
+            }, 2000);
```
Comment (juriczech) [inline suite-native/module-accounts-import/src/screens/XpubScanScreen.tsx]: nit: Is this timeout increased due to performance? Would be nice to expand on the comment so it's clear what changing the value could possibly affect

File: jest.config.base.js [hunk 11]
```diff
@@ -9,8 +9,11 @@ const babelConfig = {
             },
         ],
     ],
+    plugins: [['@babel/plugin-proposal-decorators', { version: '2023-05' }]],
 };
 
+console.log(process.cwd());
```
Comment (PeKne) [inline jest.config.base.js]: nit: forgotten console.log().

File: suite-native/accounts/src/transactionCacheMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import { TransactionCacheEngine } from '@suite-common/transaction-cache-engine';
+import { createMiddlewareWithExtraDeps } from '@suite-common/redux-utils';
+import { accountsActions } from '@suite-common/wallet-core';
+
+export const prepareTransactionCacheMiddleware = createMiddlewareWithExtraDeps(
+    (action, { next }) => {
+        if (
+            accountsActions.updateAccount.match(action) ||
+            accountsActions.createAccount.match(action)
```
Comment (PeKne) [inline suite-native/accounts/src/transactionCacheMiddleware.ts]: During discovery, we are calling `accountActions.createIndexLabeledAccount`. That one should be handler here as well.

File: suite-common/transaction-cache-engine/src/utils/balanceHistory.ts [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+const getAccountBalanceHistoryRipple = ({
+    transactions,
+}: {
+    transactions: AccountTransaction[];
+}): AccountBalanceHistory[] => {
+    const summaryMap = new Map<number, AccountBalanceHistory>();
+
+    transactions.forEach(tx => {
+        const { blockTime } = tx;
+        let sentSat = new BigNumber(0);
```
Comment (PeKne) [inline suite-common/transaction-cache-engine/src/utils/balanceHistory.ts]: nit: Ripple does not use `sats`, would be nice to change name of these variables.

File: suite-common/transaction-cache-engine/src/types.ts [hunk 14]
```diff
@@ -0,0 +1,37 @@
+import BigNumber from 'bignumber.js';
+
+import { NetworkSymbol } from '@suite-common/wallet-config';
+import { AccountTransaction } from '@trezor/connect';
+
+export type AccountUniqueKey = string & { __accountKey: 'AccountKey' };
```
Comment (PeKne) [inline suite-common/transaction-cache-engine/src/types.ts]: Could be changed to template literal type:
```suggestion
export type AccountUniqueKey = `${NetworkSymbol}-${Account["descriptor"]}` & { __accountKey: 'AccountKey' };
```

for better string type integrity.

File: suite-common/transaction-cache-engine/src/MemoryStorage.ts [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+    async getTransactions({
+        coin,
+        descriptor,
+    }: AccountUniqueParams): Promise<AccountTransaction[]> {
+        const accountUniqueKey = getAccountUniqueKey({ coin, descriptor });
+        return this.accountTransactions[accountUniqueKey] || [];
+    }
+
+    private getAccountTxids(account: AccountUniqueParams) {
```
Comment (PeKne) [inline suite-common/transaction-cache-engine/src/MemoryStorage.ts:92]: The transaction array could be pottentionally very long. What about memoizing this with `proxy-memoize`? Just an idea.

File: suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+            const numberOfExistingTransactions = await this.storage.getNumberOfExistingTransactions(
+                {
+                    coin,
+                    descriptor,
+                    txids,
+                },
+            );
+
+            // We have fetched all transactions that are not already saved in the storage.
+            if (numberOfExistingTransactions === txids.length) {
```
Comment (PeKne) [inline suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts:149]: I don't get this condition. How does the storage know, that it has all the missing transactions?

File: suite-common/graph/src/graphDataFetching.ts [hunk 17]
```diff
@@ -47,6 +51,33 @@ export const addBalanceForAccountMovementHistory = (
...
+        descriptor,
+        suppressBackupWarning: true,
+    });
+
+    if (!accountInfo?.success) {
+        throw new Error(`Get account balance info error: ${accountInfo.payload.error}`);
+    }
+
+    switch (networkType) {
+        case 'ripple':
```
Comment (PeKne) [inline suite-common/graph/src/graphDataFetching.ts:74]: comment why is this needed?

File: suite-common/graph/src/graphDataFetching.ts [hunk 18]
```diff
@@ -65,8 +96,14 @@ export const getAccountBalanceHistory = async ({
         return accountBalanceHistoryCache[cacheKey];
     }
 
-    const [accountMovementHistory, accountInfo] = await Promise.all([
-        TrezorConnect.blockchainGetAccountBalanceHistory({
+    const getBalanceHistory = async () => {
```
Comment (PeKne) [inline suite-common/graph/src/graphDataFetching.ts:100]: Two simillar functions with the same name `getBalanceHistory` are defined in both `graphDataFetching.ts` and `graphBalanceEvents.ts`. Could not it be defined outside of these files and reused?

**Review by PeKne — COMMENTED** [review]
Beautiful work with the decorators 💪.

File: suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts (hunk 8 above)
Comment (Nodonisko) [reply suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts:21]: It's for debug, so once you will have to debug that you can change it to true

File: suite-common/graph/src/graphDataFetching.ts (hunk 7 above)
Comment (Nodonisko) [reply suite-common/graph/src/graphDataFetching.ts:138]: Not latest accountBalance from getAccountInfo

File: suite-native/accounts/src/transactionCacheMiddleware.ts (hunk 9 above)
Comment (Nodonisko) [reply suite-native/accounts/src/transactionCacheMiddleware.ts]: Good catch this is not neccessary anymore

File: suite-native/module-accounts-import/src/screens/XpubScanScreen.tsx (hunk 10 above)
Comment (Nodonisko) [reply suite-native/module-accounts-import/src/screens/XpubScanScreen.tsx]: I will revert that, leftovers from debug

File: suite-native/accounts/src/transactionCacheMiddleware.ts (hunk 12 above)
Comment (Nodonisko) [reply suite-native/accounts/src/transactionCacheMiddleware.ts]: Shoudn't be necessary as `useTransactionCache` hook should handle all added accounts no matters where they come from. I removed this part of middleware.

File: suite-common/transaction-cache-engine/src/types.ts (hunk 14 above)
Comment (Nodonisko) [reply suite-common/transaction-cache-engine/src/types.ts]: that is cool

File: suite-common/transaction-cache-engine/src/MemoryStorage.ts (hunk 15 above)
Comment (Nodonisko) [reply suite-common/transaction-cache-engine/src/MemoryStorage.ts:92]: perf is not a concern right now as this will be replaced with SQL anyway

File: suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts (hunk 16 above)
Comment (Nodonisko) [reply suite-common/transaction-cache-engine/src/TransactionCacheEngine.ts:149]: Yea, this one it's pretty tricky if I see it now. I added better comment.

Basically you call `this.storage.getNumberOfExistingTransactions` on latest page of transaction fetched, if all exists you know you don't need to go to next page because you have already fetched all missing transactions.

File: suite-common/graph/src/graphDataFetching.ts (hunk 17 above)
Comment (Nodonisko) [reply suite-common/graph/src/graphDataFetching.ts:74]: I don't know either :D but I added comment at least

File: suite-common/graph/src/graphDataFetching.ts (hunk 18 above)
Comment (Nodonisko) [reply suite-common/graph/src/graphDataFetching.ts:100]: They should be replaced soon by fetching everything from TransactionCache so I don't want to invest more energy into that.

File: suite-common/transaction-cache-engine/src/utils/balanceHistory.ts (hunk 13 above)
Comment (Nodonisko) [reply suite-common/transaction-cache-engine/src/utils/balanceHistory.ts]: fixed

**Review by PeKne — APPROVED** [review]

File: suite-common/wallet-config/src/networksConfig.ts (hunk 6 above)
Comment (Nodonisko) [reply suite-common/wallet-config/src/networksConfig.ts]: done

File: suite-common/transaction-cache-engine/src/tests/TransactionCacheEngine.test.ts (hunk 5 above)
Comment (Nodonisko) [reply suite-common/transaction-cache-engine/src/tests/TransactionCacheEngine.test.ts:9]: Thanks, I will be doing follow-up where I plan to refactor tests even more so I will do it then

**Review by juriczech — APPROVED** [review]
